osu.Game.Rulesets.Osu/Difficulty/Interp/CubicInterp.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Interp
+{
+    /// <summary>
+    /// Finite difference cubic hermite interpolation
+    /// </summary>
+    public class CubicInterp
+    {
+        public static double TwoPointDerivative(double x, double val, double xNext, double valNext)
+        {
+            return (valNext - val) / (xNext - x);
+        }
+
+        public static double ThreePointDerivative(double xPrev, double valPrev, double x, double val, double xNext, double valNext)
+        {
+            return (
+                (xNext - x) * TwoPointDerivative(xPrev, valPrev, x, val)
+                + (x - xPrev) * TwoPointDerivative(x, val, xNext, valNext)
+                ) / (xNext - xPrev);
+        }
```

Comment:
I'm lost as to what is the purpose of this. Isn't this essentially tantamount to `TwoPointDerivative(xPrev, valPrev, xNext, valNext)`, but with more possibility of rounding errors?

Reply:
It's not the same, that first line is `(xNext-x) * (val-valPrev)/(x-xPrev)`. It doesn't cancel out to anything. 

The idea was to weight each side of the derivative by how close it is. If you do `TwoPointDerivative(xPrev, valPrev, xNext, valNext)` then it's kind of the same but backwards, the furthest away point gets the most weight.

Reply:
I see, I was assuming a uniform grid. Thanks for clarifying.